src/desktop_entry/mod.rs

```diff
@@ -0,0 +1,890 @@
+
+    pub fn get_type(&self) -> Result<String> {
+        let err = Error(vec!["Could not read Type".to_string()]);
+        self.type_string.clone().ok_or(err)
+    }
+
+    pub fn get_exec(&self) -> Result<String> {
+        let err = Error::from("Could not read Exec");
+        self.exec.clone().ok_or(err)
+    }
+
+    pub fn get_url(&self) -> Result<String> {
+        let err = Error::from("Could not read URL");
+        self.url.clone().ok_or(err)
+    }
+
+    fn from_hash_map(group_name: String, hashmap: &HashMap<String, String>) -> Result<Self> {
+        // String type
+        let type_string = hashmap.get("Type").cloned();
+        let version = hashmap.get("Version").cloned();
+        let exec = hashmap.get("Exec").cloned();
+        let try_exec = hashmap.get("TryExec").cloned();
+        let path = hashmap.get("Path").cloned();
+        let startup_wm_class = hashmap.get("StartupWMClass").cloned();
+        let url = hashmap.get("URL").cloned();
+        // IconString
+        let icon = hashmap.get("Icon").cloned();
+        // LocalString type
+        let name = LocaleString::from_hashmap("Name", hashmap)?;
+        let generic_name = LocaleString::from_hashmap("GenericName", hashmap)?;
+        let comment = LocaleString::from_hashmap("Comment", hashmap)?;
+        // LocaleStrings type
+        let keywords = LocaleStrings::from_hashmap("Keywords", hashmap)?;
+        // Bool type
+        let no_display = hashmap.get("NoDisplay").parse()?;
+        let hidden = hashmap.get("Hidden").parse()?;
+        let dbus_activatable = hashmap.get("DBusActivatable").parse()?;
+        let terminal = hashmap.get("Terminal").parse()?;
+        let startup_notify = hashmap.get("StartupNotify").parse()?;
+        let prefers_non_default_gpu = hashmap.get("PrefersNonDefaultGPU").parse()?;
+        // Strings type
+        let only_show_in = hashmap.get("OnlyShowIn").parse()?;
+        let not_show_in = hashmap.get("NotShowIn").parse()?;
+        let actions = hashmap.get("Actions").parse()?;
+        let mime_type = hashmap.get("MimeType").parse()?;
+        let categories = hashmap.get("Categories").parse()?;
+        let implements = hashmap.get("Implements").parse()?;
+
+        let desktop_entry = Group {
+            group_name,
+            type_string,
+            version,
+            name,
+            generic_name,
+            no_display,
+            comment,
+            icon,
+            hidden,
+            only_show_in,
+            not_show_in,
+            dbus_activatable,
+            try_exec,
+            exec,
+            path,
+            terminal,
+            actions,
+            mime_type,
+            categories,
+            implements,
+            keywords,
+            startup_notify,
+            startup_wm_class,
+            url,
+            prefers_non_default_gpu,
+        };
+
+        desktop_entry.validate()?;
+        Ok(desktop_entry)
+    }
+
+    fn check_not_show_in(&self) -> Result<()> {
```

Comment:
[Registered OnlyShowIn Environments](https://specifications.freedesktop.org/menu-spec/latest/apb.html) just list some well known, but not the all *valid*.

As [Recognized desktop entry keys](https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s06.html) say, 
> `$XDG_CURRENT_DESKTOP` should have been set by the login manager, according to the value of the DesktopNames found in the session file

So this check is not necessary.

Same to `check_only_show_in`